audio/drivers/switch_libnx_audren_audio.c

```diff
@@ -0,0 +1,357 @@
+/*  RetroArch - A frontend for libretro.
+ *  Copyright (C) 2019      - p-sam
+ *
+ *  RetroArch is free software: you can redistribute it and/or modify it under the terms
+ *  of the GNU General Public License as published by the Free Software Found-
+ *  ation, either version 3 of the License, or (at your option) any later version.
+ *
+ *  RetroArch is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ *  without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
+ *  PURPOSE.  See the GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along with RetroArch.
+ *  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <malloc.h>
+#include <stdint.h>
+
+#include <switch.h>
+
+#include "../../retroarch.h"
+#include "../../verbosity.h"
+
+#define BUFFER_COUNT 5
+
+static const int sample_rate           = 48000;
+static const int num_channels          = 2;
+static const uint8_t sink_channels[]   = { 0, 1 };
+
+static const AudioRendererConfig audio_renderer_config =
+{
+   .output_rate     = AudioRendererOutputRate_48kHz,
+   .num_voices      = 24,
+   .num_effects     = 0,
+   .num_sinks       = 1,
+   .num_mix_objs    = 1,
+   .num_mix_buffers = 2,
+};
+
+typedef struct {
+   AudioDriver drv;
+   void* mempool;
+   AudioDriverWaveBuf wavebufs[BUFFER_COUNT];
+   AudioDriverWaveBuf* current_wavebuf;
+   void* current_pool_ptr;
+   size_t current_size;
+   size_t buffer_size;
+   size_t samples;
+   Mutex update_lock;
+   bool nonblocking;
+} libnx_audren_t;
+
+static void *libnx_audren_audio_init(const char *device, unsigned rate, unsigned latency,
+      unsigned block_frames,
+      unsigned *new_rate)
+{
+   unsigned i, j;
+   libnx_audren_t *aud;
+   Result rc;
+   int mpid;
+   size_t mempool_size;
+   unsigned real_latency;
+
+   RARCH_LOG("[Audio]: Using libnx_audren driver\n");
+
+   aud = (libnx_audren_t*)calloc(1, sizeof(libnx_audren_t));
+
+   if (!aud)
+   {
+      RARCH_ERR("[Audio]: struct alloc failed\n");
+      goto fail;
+   }
+
+   real_latency = MAX(5, latency);
```

Comment:
I propose to change this to
```real_latency = MAX(20, latency);```
since audren device starts to stutter with real (measured) latency values <20 ms. Note the corresponding division by four in samples calculation (see other comment).